src/main/java/net/fabricmc/loom/providers/MappingsProvider.java

```diff
@@ -66,50 +112,133 @@ public void provide(DependencyInfo dependency, Project project, LoomGradleExtens
 		project.getLogger().lifecycle(":setting up mappings (" + dependency.getDependency().getName() + " " + dependency.getResolvedVersion() + ")");
 
 		String version = dependency.getResolvedVersion();
-		File mappingsJar = dependency.resolveFile().orElseThrow(() -> new RuntimeException("Could not find dependency " + dependency));
+		File mappingsJar = dependency.resolveFile().orElseThrow(() -> new RuntimeException("Could not find yarn mappings: " + dependency));
+
+		// Hacky but seems like the easiest way to get the classifier
+		this.isV2 = mappingsJar.getName().endsWith("-v2.jar");
 
-		this.mappingsName = dependency.getDependency().getGroup() + "." + dependency.getDependency().getName();
+		this.mappingsName = StringUtils.removeSuffix(dependency.getDependency().getGroup() + "." + dependency.getDependency().getName(), "-unmerged");
+		if (this.isV2) mappingsName += "-v2";
 
 		Version mappingsVersion = new Version(version);
 		this.minecraftVersion = mappingsVersion.getMinecraftVersion();
 		this.mappingsVersion = mappingsVersion.getMappingsVersion();
 
 		initFiles(project);
 
-		if (!MAPPINGS_DIR.exists()) {
-			MAPPINGS_DIR.mkdir();
+		Files.createDirectories(mappingsDir);
+		Files.createDirectories(mappingsStepsDir);
+
+		if (!tinyMappings.exists()) {
+			storeMappings(project, minecraftProvider, mappingsJar);
 		}
 
-		if (!MAPPINGS_TINY_BASE.exists() || !MAPPINGS_TINY.exists()) {
-			if (!MAPPINGS_TINY_BASE.exists()) {
-				project.getLogger().lifecycle(":extracting " + mappingsJar.getName());
+		mappedProvider = new MinecraftMappedProvider();
+		mappedProvider.initFiles(project, minecraftProvider, this);
+		mappedProvider.provide(dependency, project, extension, postPopulationScheduler);
+	}
 
-				try (FileSystem fileSystem = FileSystems.newFileSystem(mappingsJar.toPath(), null)) {
-					Path fileToExtract = fileSystem.getPath("mappings/mappings.tiny");
-					Files.copy(fileToExtract, MAPPINGS_TINY_BASE.toPath());
-				}
-			}
+	private void storeMappings(Project project, MinecraftProvider minecraftProvider, File yarnJar) throws IOException {
+		if (isV2) {
+			// These are unmerged v2 mappings
+			Dependency intermediaryDep = project.getDependencies().add("compile", String.format("net.fabricmc:intermediary:%s:v2", minecraftProvider.minecraftVersion));
```

Comment:
Are the yarn names also added to the classpath? as having 2 jars that have the same 2 file names in on the CP might be annoying.

Reply:
Well, the mappings file itself doesn't get added to the classpath, just the jar that contains them.

Reply:
Yeah, I know that. Does the yarn jar also get added (I guess it must as its needed by loader)

Reply:
The combined yarn-intermediary jar is what is added to the runtime classpath that loader uses.

Reply:
Ok, so then this is definitely wrong then

Reply:
the `"compile"` part? or adding it as a dependency alltogether?

Reply:
`compile` is not right all togeather. If you still want to use gradle's maven downloading you could add it to its own config.

What I would do here is just download it directally tbh. I dont see any reason not to?

Reply:
sure we'll do that